diff --git a/src/video/windows/SDL_windowsevents.c b/src/video/windows/SDL_windowsevents.c
index 8b1ee1cfc..1cab80c73 100644
--- a/src/video/windows/SDL_windowsevents.c
+++ b/src/video/windows/SDL_windowsevents.c
@@ -375,9 +375,8 @@ static void WIN_UpdateFocus(SDL_Window *window, bool expect_focus, DWORD pos)
         WIN_CheckAsyncMouseRelease(WIN_GetEventTimestamp(), data);
         WIN_UpdateClipCursor(window);
 
-        /*
-         * FIXME: Update keyboard state
-         */
+        WIN_UpdateModifiers();
+
         WIN_CheckClipboardUpdate(data->videodata);
 
         SDL_ToggleModState(SDL_KMOD_CAPS, (GetKeyState(VK_CAPITAL) & 0x0001) ? true : false);
@@ -1568,6 +1567,8 @@ LRESULT CALLBACK WIN_WindowProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lPara
             break;
         }
 
+        WIN_UpdateModifiers();
+
         bool virtual_key = false;
         Uint16 rawcode = 0;
         SDL_Scancode code = WindowsScanCodeToSDLScanCode(lParam, wParam, &rawcode, &virtual_key);
diff --git a/src/video/windows/SDL_windowskeyboard.c b/src/video/windows/SDL_windowskeyboard.c
index b114772d8..d2b7d9d7f 100644
--- a/src/video/windows/SDL_windowskeyboard.c
+++ b/src/video/windows/SDL_windowskeyboard.c
@@ -247,6 +247,26 @@ void WIN_QuitKeyboard(SDL_VideoDevice *_this)
     keymap_cache_size = 0;
 }
 
+static void WIN_HandleModifier(SDL_Scancode code, int vk)
+{
+    bool pressed = (GetAsyncKeyState(vk) & 0x8000) != 0;
+    if (pressed != SDL_GetKeyboardState(NULL)[code]) {
+        SDL_SendKeyboardKey(0, SDL_GLOBAL_KEYBOARD_ID, 0, code, pressed);
+    }
+}
+
+void WIN_UpdateModifiers(void)
+{
+    WIN_HandleModifier(SDL_SCANCODE_LSHIFT, VK_LSHIFT);
+    WIN_HandleModifier(SDL_SCANCODE_RSHIFT, VK_RSHIFT);
+    WIN_HandleModifier(SDL_SCANCODE_LCTRL, VK_LCONTROL);
+    WIN_HandleModifier(SDL_SCANCODE_RCTRL, VK_RCONTROL);
+    WIN_HandleModifier(SDL_SCANCODE_LALT, VK_LMENU);
+    WIN_HandleModifier(SDL_SCANCODE_RALT, VK_RMENU);
+    WIN_HandleModifier(SDL_SCANCODE_LGUI, VK_LWIN);
+    WIN_HandleModifier(SDL_SCANCODE_RGUI, VK_RWIN);
+}
+
 void WIN_ResetDeadKeys(void)
 {
     /*
diff --git a/src/video/windows/SDL_windowskeyboard.h b/src/video/windows/SDL_windowskeyboard.h
index ec0e91d25..8c9e13700 100644
--- a/src/video/windows/SDL_windowskeyboard.h
+++ b/src/video/windows/SDL_windowskeyboard.h
@@ -26,6 +26,7 @@
 extern void WIN_InitKeyboard(SDL_VideoDevice *_this);
 extern void WIN_UpdateKeymap(bool send_event);
 extern void WIN_QuitKeyboard(SDL_VideoDevice *_this);
+extern void WIN_UpdateModifiers(void);
 
 extern void WIN_ResetDeadKeys(void);
 
